_posts/2020-07-09-implementing-retry-with-resilience4j.md

```diff
@@ -0,0 +1,386 @@
+                                    maxAttempts(6).
+                                    intervalFunction(IntervalFunction.ofExponentialBackoff(1000, 2)).
+                                    build();
+```
+
+Sample output below shows the exponentially increasing attempt timestamps:
+
+```
+Searching for flights; current time = 20:37:02 684
+Operation failed
+Searching for flights; current time = 20:37:03 727
+Operation failed
+Searching for flights; current time = 20:37:05 731
+Operation failed
+Searching for flights; current time = 20:37:09 731
+Operation failed
+Searching for flights; current time = 20:37:17 731
+```
+
+`IntervalFunction` also provides an `exponentialRandomBackoff` method which combines both the approaches above. We can also provide custom implementations of `IntervalFunction`.
+
+### Retrying Asynchronous Operations
+
+The examples we saw till now were all synchronous calls. Let's see how to retry asynchronous operations. Suppose we were searching for flights asynchronously like this:
+
+```java
+CompletableFuture.supplyAsync(() -> service.searchFlights(request)).thenAccept(System.out::println);
+```
+
+The `searchFlight` call happens on a different thread and when it returns, the returned `List<Flight>` is passed to `thenAccept` which just prints it.
+
+We can do retries for asynchronous operations like above using the `executeCompletionStage` method on the `Retry` object. This method takes two parameters - a `ScheduledExecutorService` on which the retry will be scheduled  and a `Supplier<CompletionStage>` that will be decorated. It decorates and executes the `CompletionStage` and then returns a `CompletionStage` on which we can call `thenAccept` as before:
+
+```java
+ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor();
+Supplier<CompletionStage<List<Flight>>> completionStageSupplier = () -> CompletableFuture.supplyAsync(() -> service.searchFlights(request));
+retry.executeCompletionStage(scheduler, completionStageSupplier).thenAccept(System.out::println);
+
+```
+
+In a real application, we would use a shared thread pool (`Executors.newScheduledThreadPool`) for scheduling the retries instead of the single-threaded scheduled executor shown here. 
+
+### Retry Events
+
+In all these examples, the decorator has been a black box - we don't know when an attempt failed and the framework code is attempting a retry. Suppose for a given request, we wanted to log some details like the attempt count or the wait time until the next attempt. We can do these using Retry events that are published at different points of execution. `Retry` has an `EventPublisher` that has methods like `onRetry`, `onSuccess`, etc. We can collect and log details by using these:
+
+```java
+Retry.EventPublisher publisher = retry.getEventPublisher();
+publisher.onRetry(event -> System.out.println(event.toString()));
+publisher.onSuccess(event -> System.out.println(event.toString()));
+```
+
+Sample output:
```

Comment:
I removed the sample output here, because the code makes it easy to grasp already.